File: 백준/Silver/11659. 구간 합 구하기 4/구간 합 구하기 4.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

class Main {
    static StringTokenizer st;
    static int[] seq;
    static int n, m;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        st = new StringTokenizer(br.readLine());
        n = nextInt();
        m = nextInt();

        st = new StringTokenizer(br.readLine());
        seq = new int[n + 1];

        for (int i = 1; i <= n; i++) {
            seq[i] = seq[i - 1] + nextInt();
        }

        StringBuilder sb = new StringBuilder();

        while (m-- > 0) {
            st = new StringTokenizer(br.readLine());
            int s = nextInt();
            int e = nextInt();

            sb.append(seq[e] - seq[s - 1]).append('\n');
        }

        System.out.println(sb);
        br.close();
    }

    static int nextInt() {
        return Integer.parseInt(st.nextToken());
    }
}

File: 백준/Silver/11659. 구간 합 구하기 4/구간 합 구하기 4.kt
import java.util.*

fun main() = with(System.`in`.bufferedReader()) {
    val (n, m) = StringTokenizer(readLine()).run { nextToken().toInt() to nextToken().toInt() }
    val seg = SegmentTree(n)

    with(StringTokenizer(readLine())) {
        repeat(n) { seg.update(it, nextToken().toInt()) }
    }

    val bw = System.out.bufferedWriter()

    repeat(m) {
        val (s, e) = StringTokenizer(readLine()).run { nextToken().toInt() - 1 to nextToken().toInt() - 1 }
        bw.append("${seg.query(s, e)}\n")
    }
    close()

    bw.flush()
    bw.close()
}

class SegmentTree(n: Int) {
    val tree: IntArray
    var cap = 1

    init {
        while (cap < n) cap *= 2
        tree = IntArray(cap shl 1)
    }

    fun update(i: Int, v: Int) {
        var node = cap + i

        while (node > 0) {
            tree[node] += v
            node = node ushr 1
        }
    }

    fun query(qs: Int, qe: Int): Int {
        var ret = 0
        var s = cap + qs
        var e = cap + qe

        while (s < e) {
            if (s and 1 == 1) ret += tree[s++]
            if (e and 1 == 0) ret += tree[e--]

            s = s ushr 1
            e = e ushr 1
        }
        if (s == e) ret += tree[s]

        return ret
    }
}